src/anomalib/engine/engine.py

```diff
@@ -0,0 +1,189 @@
+
+log = logging.getLogger(__name__)
+
+
+class UnassignedException(Exception):
+    ...
+
+
+class _Cache:
+    """Cache arguments.
+
+    Args:
+        (**kwargs): Trainer arguments that are cached
+
+    Example:
+        >>> cache = _Cache(max_epochs=100, devices=0)
+        >>> model = Padim()
+        >>> cache.update(model)
+        >>> Trainer(**cache.params)
+    """
+
+    def __init__(self, **kwargs):
+        self._cache = {**kwargs}
+
+    def update(self, model: AnomalyModule):
+        """Replace cached arguments with arguments retrieved from the model.
+
+        Args:
+            model (AnomalyModule): The model used for training
+        """
+        for key, value in model.trainer_arguments.items():
+            if key in self._cache:
+                if self._cache[key] != value:
+                    log.info(f"Overriding {key} from {self._cache[key]} with {value} for {model.__class__.__name__}")
+                self._cache[key] = value
+
+    def needs_override(self, model: AnomalyModule) -> bool:
+        for key, value in model.trainer_arguments.items():
+            if key in self._cache and self._cache[key] != value:
+                return True
+        return False
+
+    @property
+    def params(self) -> dict[str, Any]:
+        return self._cache
+
+
+class Engine:
+    """Anomalib Engine.
+
+    Note:
+        Refer to PyTorch Lightning's Trainer for a list of parameters for details on other Trainer parameters.
+
+    Args:
+        callbacks: Add a callback or list of callbacks.
+    """
+
+    def __init__(
+        self,
+        callbacks: list[Callback] = [],
+        normalizer: NormalizationMethod | DictConfig | Callback | str = NormalizationMethod.MIN_MAX,
+        threshold: BaseThreshold
+        | tuple[BaseThreshold, BaseThreshold]
+        | DictConfig
+        | ListConfig
+        | str = F1AdaptiveThreshold(),
+        task: TaskType = TaskType.SEGMENTATION,
+        image_metrics: list[str] | None = None,
+        pixel_metrics: list[str] | None = None,
+        visualization: DictConfig | None = None,
+        **kwargs,
+    ) -> None:
+        self._cache = _Cache(callbacks=[*callbacks, RichProgressBar()], **kwargs)
+        self.normalizer = normalizer
+        self.threshold = threshold
+        self.task = task
+        self.image_metric_names = image_metrics
+        self.pixel_metric_names = pixel_metrics
+        self.visualization = visualization
+
+        self.lightning_module: AnomalyModule
+        self._trainer: Trainer | None = None
+
+    @property
+    def trainer(self):
+        if not self._trainer:
+            msg = "``self.trainer`` is not assigned yet."
+            raise UnassignedException(msg)
+        return self._trainer
+
+    def _setup_trainer(self, model: AnomalyModule):
+        """Instantiates the trainer based on the model parameters."""
+        if self._cache.needs_override(model) or self._trainer is None:
```

Comment:
Do we need this method? Looks like we are going over this list anyway. Maybe we could just call `.update` directly. If it requires update, we could override the args. Otherwise, they could stay the same. 

Thoughts?

Reply:
This also ensures that the new trainer is created only when the arguments are updated. Maybe there might be a better way to write this method.